File: MyProject/Source/MyProject/Private/BasicFrame/Libs/FrameHandle/FrameTimerMgr.cpp
#include "MyProject.h"
#include "FrameTimerMgr.h"
#include "IDelayHandleItem.h"
#include "FrameTimerItem.h"
#include "UtilContainers.h"

FrameTimerMgr::FrameTimerMgr()
{

}

void FrameTimerMgr::addObject(IDelayHandleItem* delayObject, float priority)
{
	// 检查当前是否已经在队列中
	FrameTimerItem* frameTimerItem = (FrameTimerItem*)delayObject;
	if (UtilVector::IndexOf(mTimerList, frameTimerItem) == -1)
	{
		if (isInDepth())
		{
			DelayHandleMgrBase::addObject(delayObject, priority);
		}
		else
		{
			mTimerList.push_back((FrameTimerItem*)delayObject);
		}
	}
}

void FrameTimerMgr::delObject(IDelayHandleItem* delayObject)
{
	// 检查当前是否在队列中
	FrameTimerItem* frameTimerItem = (FrameTimerItem*)delayObject;
	if (UtilVector::IndexOf(mTimerList, frameTimerItem) != -1)
	{
		((FrameTimerItem*)delayObject)->mIsDisposed = true;
		if (isInDepth())
		{
			DelayHandleMgrBase::addObject(delayObject);
		}
		else
		{
			for(FrameTimerItem* item : mTimerList)
			{
				if (item == delayObject)
				{
					UtilVector::Remove(mTimerList, item);
					break;
				}
			}
		}
	}
}

void FrameTimerMgr::Advance(float delta)
{
	incDepth();

	for(FrameTimerItem* timerItem : mTimerList)
	{
		if (!timerItem->isClientDispose())
		{
			timerItem->OnFrameTimer();
		}
		if (timerItem->mIsDisposed)
		{
			delObject(timerItem);
		}
	}

	decDepth();
}

File: MyProject/Content/MyAsset/Lua/MyLua/Libs/FrameHandle/FrameTimerMgr.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.DataStruct.MList");
MLoader("MyLua.Libs.DelayHandle.DelayHandleMgrBase");

local M = GlobalNS.Class(GlobalNS.DelayHandleMgrBase);
M.clsName = "FrameTimerMgr";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mTimerList = GlobalNS.new(GlobalNS.MList);
    self.mDelList = GlobalNS.new(GlobalNS.MList);
end

function M:addObject(delayObject, priority)
    -- 检查当前是否已经在队列中
    if self.mTimerList:IndexOf(delayObject) == -1 then
        if self:bInDepth() then
            M.super.addObject(self, delayObject, priority);
        else
            self.mTimerList:Add(delayObject);
        end
    end
end

function M:removeObject(delayObject)
    -- 检查当前是否在队列中
    if not self.mTimerList:IndexOf(delayObject) == -1 then
        delayObject.mIsDisposed = true;
        if self:bInDepth() then
            M.super.addObject(self, delayObject);
        else
            for key, item in ipairs(self.mTimerList.list()) do
                if item == delayObject then
                    self.mTimerList:Remove(item);
                    break;
                end
            end
        end
    end
end

function M:Advance(delta)
    self.incDepth();

    for key, timerItem in ipairs(self.mTimerList.list()) do
        if not timerItem:getClientDispose() then
            timerItem:OnFrameTimer();
        end
        if timerItem.mIsDisposed then
            self:removeObject(timerItem);
        end
    end

    self:decDepth();
end

return M;